tools/topology/m4/dai.m4

```diff
@@ -147,16 +147,39 @@ define(`D_DAI', `SectionDAI."'N_DAI`" {'
 `	capture "'$3`"'
 `}')
 
-dnl DAI Config)
-define(`N_DAI_CONFIG', `DAICONFIG.'$1)
-dnl DAI_CONFIG(type, idx, link_id, name, sai_config/esai_config/ssp_config/dmic_config)
-define(`DO_DAI_CONFIG',
-`SectionHWConfig."'$1$2`" {'
+define(`DO_HW_CONFIG', `ifelse(`$#', `3', , `$#', `4',
+`SectionHWConfig."'$1$2_$3`" {'
+`'
+`	id		"'$3`"'
+`ifelse($1, `DMIC', `}')'
+`'
+`	$4',
+`SectionHWConfig."'$1$2_$3`" {'
 `'
 `	id		"'$3`"'
+`ifelse($1, `DMIC', `}')'
 `'
-`	ifelse($1, `SSP', $5, $1, `HDA', $5, $1, `ALH', $5, $1, `ESAI', $5, $1, `SAI', $5, `}')'
-`ifelse($1, `DMIC', $5, `')'
+`	$4'
```

Comment:
@juimonen this is very impressive but I must say it is very hard to review. I have a simpler suggestion. Why not have a new macro called DAI_CONFIG_MULTI() which takes a list of SSP_CONFIG() instead of the vanilla DAI_CONFIG() which takes only one. This way you dont have to worry about breaking existing topologies.

Reply:
yes actually very good idea... m4 makes me be blind quite often...